include/gatekeeper_sol.h

```diff
@@ -0,0 +1,122 @@
+ */
+
+#ifndef _GATEKEEPER_SOL_H_
+#define _GATEKEEPER_SOL_H_
+
+#include <rte_approx.h>
+#include <rte_cycles.h>
+#include <rte_reciprocal.h>
+
+#include "list.h"
+
+struct priority_req {
+	/* Doubly-linked list node. */
+	struct list_head    list;
+	/* The packet for this request. */
+	struct rte_mbuf     *pkt;
+	/* The priority of this request. */
+	uint8_t             priority;
+};
+
+/* Number of priorities available in the IP DSCP field. */
+#define GK_NUM_REQ_PRIORITIES (64)
+
+/*
+ * XXX The DPDK packet scheduler uses __rte_cache_aligned
+ * on member @memory and on the struct as a whole. Should
+ * it be used here?
+ */
+struct req_queue {
+	/* Length of the priority queue. */
+	uint32_t              len;
+	/* Number of possible priorities in the priority queue. */
+	uint16_t              num_priorities;
+	/* The highest priority of any packet currently in the queue. */
+	uint16_t              highest_priority;
+	/* The lowest priority of any packet currently in the queue. */
+	uint16_t              lowest_priority;
+
+	/* Token bucket algorithm state. */
+
+	/* Token bucket time measured in bytes. */
+	uint64_t              tb_time;
+	/* Length of a token bucket period. */
+	uint32_t              tb_period;
+	/* Number of credits accrued for every token bucket period. */
+	uint32_t              tb_credits_per_period;
+	/* Capacity of the token bucket (the max number of credits). */
+	uint32_t              tb_size;
+	/* Number of credits currently in the token bucket. */
+	uint32_t              tb_credits;
+
+	/* Current CPU time measured in CPU cyles. */
+	uint64_t              time_cpu_cycles;
+	/* Current CPU time measured in bytes. */
+	uint64_t              time_cpu_bytes;
+	/* Current NIC TX time measured in bytes. */
+	uint64_t              time;
+	/* CPU cycles per byte. */
+	struct rte_reciprocal inv_cycles_per_byte;
+
+	/*
+	 * The head of the priority queue, referencing the node
+	 * that contains the packet with the highest priority.
+	 */
+	struct list_head      head;
+	/* Array of pointers to packets of each priority. */
+	struct priority_req   *priorities[GK_NUM_REQ_PRIORITIES];
+};
+
+/* Configuration for the Solicitor functional block. */
+struct sol_config {
+	unsigned int       lcore_id;
+
+	/* Maximum number of requests to store in priority queue at once. */
+	unsigned int       pri_req_max_len;
+
+	/* Maximum request enqueue/dequeue size. */
+	unsigned int       enq_burst_size;
+	unsigned int       deq_burst_size;
+
+	/*
+	 * The fields below are for internal use.
+	 * Configuration files should not refer to them.
+	 */
+
+	/* Priority queue for request packets. */
+	struct req_queue   req_queue;
+
+	/*
+	 * Mailbox into which GK instances enqueue request packets
+	 * to be serviced and sent out by the Solicitor.
+	 */
+	struct mailbox     mb;
+
+	/* TX queue on the back interface. */
+	uint16_t           tx_queue_back;
+	struct net_config  *net;
+	struct gk_config   *gk;
```

Comment:
The field `gk` is never used. Moreover, it's not the Solicitor block that depends on the GK blocks, but the other way around. The correct dependency should be reflected in the configuration files because if a GK block is running and the Solicitor is not, the code will crash.